File: app/src/main/java/com/example/jianming/myapplication/PicAlbumListActivity.java
package com.example.jianming.myapplication;

import android.app.Activity;
import android.content.ComponentName;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.Bundle;
import android.os.IBinder;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ListView;
import android.widget.TextView;

//import com.example.jianming.Tasks.DownloadPicListTask;
import com.example.jianming.Utils.EnvArgs;
import com.example.jianming.Utils.FileUtil;
import com.example.jianming.Utils.NetworkUtil;
import com.example.jianming.beans.PicAlbumBean;
import com.example.jianming.listAdapters.PicAlbumListAdapter;
import com.example.jianming.services.DownloadService;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.util.List;

import butterknife.Bind;
import butterknife.ButterKnife;
import butterknife.OnItemClick;


public class PicAlbumListActivity extends AppCompatActivity {


    DownloadService downLoadService = null;

    Boolean isBound = false;

    private ServiceConnection conn = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
            Log.d(TAG, "onServiceConnected");
            isBound = true;
            downLoadService = ((DownloadService.LocalBinder) service).getService();
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {
            Log.d(TAG, "onServiceDisconnected");
            downLoadService = null;
            isBound = false;
        }
    };

    public void doPicListDownloadComplete(String dirName, int index) {
        PicAlbumBean.setExistByIndex(index, 1);
        Intent intent = new Intent(this, PicAlbumActivity.class);
        intent.putExtra("name", dirName);
        intent.putExtra("index", index);
        picAlbumListAdapter.notifyDataSetChanged();
//        startActivity(intent);
    }

    PicAlbumListAdapter picAlbumListAdapter;

    Activity self = this;
    private final static String TAG = "PicListActivity";

    @Bind(R.id.list_view1)
    public ListView listView;

    @Bind(R.id.tl_custom)
    public Toolbar toolbar;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.acivity_pic_album_list);
        ButterKnife.bind(this);

        setSupportActionBar(toolbar);
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setHomeButtonEnabled(true);
            actionBar.setDisplayHomeAsUpEnabled(true);
        }



        toolbar.setNavigationOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                finish();
            }
        });
        List<PicAlbumBean> dataArray = getDataSourceFromJsonFile();
        picAlbumListAdapter = new PicAlbumListAdapter(this);
        picAlbumListAdapter.setDataArray(dataArray);
        listView.setAdapter(picAlbumListAdapter);

        bindService(new Intent(this, DownloadService.class), conn, BIND_AUTO_CREATE);

    }

    @Override
    protected void onDestroy() {
        Log.d(TAG, "onDestroy");
        super.onDestroy();
        unbindService(conn);
        isBound = false;
    }

    @OnItemClick(R.id.list_view1)
    public void doItemClick(AdapterView<?> parent, View view, int position, long id) {
        PicAlbumListAdapter.ViewHolder holder = (PicAlbumListAdapter.ViewHolder) view.getTag();
        final String name = ((TextView) view.findViewById(R.id.pic_text_view))
                .getText()
                .toString();
        final int index = holder.index;
        if (holder.exist) {
            Log.i(TAG, "you click " + index + "th item, name = " + name);
            Intent intent = new Intent(self, PicAlbumActivity.class);
            intent.putExtra("name", name);
            intent.putExtra("index", index);
            self.startActivity(intent);
        } else {
            File file = FileUtil.getAlbumStorageDir(PicAlbumListActivity.this, name);
            if (file.mkdirs()) {
                Log.i(TAG, file.getAbsolutePath() + " made");
            }
            String url = ("http://%serverIP:%serverPort/local1000/picContentAjax?id=" + index)
                    .replace("%serverIP", EnvArgs.serverIP)
                    .replace("%serverPort", EnvArgs.serverPort);
//            DownloadPicListTask.executeDownloadAlbumInfo(
//                    self,
//                    index,
//                    name,
//                    holder.downloadProcessView,
//                    url
//
//            );
            downLoadService.startDownload(index, name, holder.downloadProcessView, url);
        }
    }


    private String readIndexFile() {
        File directory = FileUtil.getAlbumStorageDir(PicAlbumListActivity.this, "file");
        File file = new File(directory, "index.json");
        String fileContent = "";
        FileInputStream fileInputStream;
        try {
            fileInputStream = new FileInputStream(file);

            /*********** another way of reading file here  ****************
             * byte[] buff = new byte[fileInputStream.available()];
             *
             * fileInputStream.read(buff);
             * fileContent = new String(buff);
             ***************************************************************/

            byte[] buff = new byte[30];
            ByteArrayOutputStream out = new ByteArrayOutputStream(30);
            int readLen;
            do {
                readLen = fileInputStream.read(buff);
                if (readLen > 0) {
                    out.write(buff, 0, readLen);
                }
            } while(readLen != -1);
            fileContent = new String(out.toByteArray());
//            Log.i("readFile", fileContent);
        } catch (IOException e) {
            e.printStackTrace();
        }
        return fileContent;
    }

    private List<PicAlbumBean> getDataSourceFromJsonFile() {
        if (isNotExistItemShown && NetworkUtil.isNetworkAvailable(this)) {
            return PicAlbumBean.getAll();
        }
        else {
            return PicAlbumBean.getAllExist();
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_album_list, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (NetworkUtil.isNetworkAvailable(this)) {
            int id = item.getItemId();

            if (id == R.id.hide_not_exist) {
                if (isNotExistItemShown) {
                    item.setTitle(R.string.show_not_exist_item);
                    isNotExistItemShown = false;
                } else {
                    item.setTitle(R.string.hide_not_exist_item);
                    isNotExistItemShown = true;
                }
                List<PicAlbumBean> dataArray = getDataSourceFromJsonFile();
                picAlbumListAdapter.setDataArray(dataArray);
                picAlbumListAdapter.notifyDataSetChanged();
            } else if (id == R.id.call_download) {
                if (isBound && downLoadService != null) {
                    downLoadService.startDownload();
                }
            }
        }

        return super.onOptionsItemSelected(item);
    }

    private boolean isNotExistItemShown = true;
}

File: app/src/main/java/com/example/jianming/myapplication/PicAlbumListActivity.kt
package com.example.jianming.myapplication

import SERVER_IP
import SERVER_PORT
import android.annotation.SuppressLint
import android.content.ComponentName
import android.content.Intent
import android.content.ServiceConnection
import android.os.Bundle
import android.os.IBinder
import android.util.Log
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.room.Room.databaseBuilder
import com.example.jianming.Tasks.ConcurrencyJsonApiTask
import com.example.jianming.util.AppDataBase
import com.example.jianming.util.NetworkUtil
import com.example.jianming.util.TimeUtil
import com.example.jianming.beans.PicAlbumBean
import com.example.jianming.beans.PicAlbumData
import com.example.jianming.dao.PicAlbumDao
import com.example.jianming.dao.PicInfoDao
import com.example.jianming.dao.UpdataStampDao
import com.example.jianming.listAdapters.PicAlbumListAdapter
import com.example.jianming.services.DownloadService
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.launch
import org.nanjing.knightingal.processerlib.RefreshListener

class PicAlbumListActivity : AppCompatActivity(), RefreshListener {


    private val TAG = "PicAlbumListActivityMD"
    private lateinit var db: AppDataBase

    private lateinit var picAlbumDao: PicAlbumDao

    private lateinit var picInfoDao: PicInfoDao

    private lateinit var updataStampDao: UpdataStampDao


    private lateinit var picAlbumListAdapter: PicAlbumListAdapter

    private lateinit var listView: RecyclerView

    private var picAlbumDataList: MutableList<PicAlbumData> = mutableListOf();

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val offline = intent.getBooleanExtra("offline", false)
        isNotExistItemShown = !offline

        db = databaseBuilder(
            applicationContext,
            AppDataBase::class.java, "database-flow1000"
        ).allowMainThreadQueries().build()

        picAlbumDao = db.picAlbumDao()
        picInfoDao = db.picInfoDao()
        updataStampDao = db.updateStampDao()

        setContentView(R.layout.activity_pic_album_list_activity_md)


        listView = findViewById(R.id.list_view11)
        listView.setHasFixedSize(true)
        val mLayoutManager: RecyclerView.LayoutManager = LinearLayoutManager(this)
        listView.layoutManager = mLayoutManager

        picAlbumListAdapter = PicAlbumListAdapter(this)
        picAlbumListAdapter.setDataArray(picAlbumDataList)
        listView.adapter = picAlbumListAdapter


    }

    var downLoadService: DownloadService? = null


    var isBound = false
    private val conn: ServiceConnection = object : ServiceConnection {
        override fun onServiceConnected(name: ComponentName, binder: IBinder) {
            Log.d(TAG, "onServiceConnected")
            isBound = true
            downLoadService = (binder as DownloadService.LocalBinder).getService()
            downLoadService!!.setRefreshListener(
                this@PicAlbumListActivity
            )
            if (isNotExistItemShown && NetworkUtil.isNetworkAvailable(this@PicAlbumListActivity)) {
                startDownloadPicIndex()
            } else {
                refreshFrontPage.invoke()
            }
        }

        override fun onServiceDisconnected(name: ComponentName) {
            Log.d(TAG, "onServiceDisconnected")
            downLoadService = null
            isBound = false
        }
    }

    override fun onPause() {
        super.onPause()
        downLoadService?.removeRefreshListener()
        downLoadService = null
        unbindService(conn)
    }

    override fun onStart() {
        super.onStart()
        bindService(Intent(this, DownloadService::class.java), conn, BIND_AUTO_CREATE)
    }

    private fun startDownloadPicIndex() {
        val updateStamp = updataStampDao.getUpdateStampByTableName("PIC_ALBUM_BEAN")
        val stringUrl = "http://${SERVER_IP}:${SERVER_PORT}/local1000/picIndexAjax?time_stamp=${updateStamp.updateStamp}"
        Log.d("startDownloadWebPage", stringUrl)
        ConcurrencyJsonApiTask.startDownload(stringUrl) { allBody ->
            val mapper = jacksonObjectMapper()
            db.runInTransaction() {
                val updateStamp = updataStampDao.getUpdateStampByTableName("PIC_ALBUM_BEAN")
                updateStamp.updateStamp = TimeUtil.currentTimeFormat()
                updataStampDao.update(updateStamp)
                val picAlbumBeanList: List<PicAlbumBean> = mapper.readValue(allBody)
                picAlbumBeanList.forEach { picAlbumDao.insert(it) }
            }

            refreshFrontPage.invoke()

            val pendingUrl = "http://${SERVER_IP}:${SERVER_PORT}/local1000/picIndexAjax?client_status=PENDING"
            ConcurrencyJsonApiTask.startDownload(pendingUrl) { pendingBody ->
                val picAlbumBeanList: List<PicAlbumBean> = mapper.readValue(pendingBody)
                if (picAlbumBeanList.isNotEmpty()) {
                    picAlbumBeanList.forEach {
                        picAlbumDao.update(it)
                        asyncStartDownload(it.id, picAlbumDataList.indexOf(picAlbumDataList.stream().filter { item ->
                            item.picAlbumBean.id == it.id
                        }.findFirst().get()));
                    }
                }
            }
            val localUrl = "http://${SERVER_IP}:${SERVER_PORT}/local1000/picIndexAjax?client_status=LOCAL"
            ConcurrencyJsonApiTask.startDownload(localUrl) { pendingBody ->
                val picAlbumBeanList: List<PicAlbumBean> = mapper.readValue(pendingBody)
                if (picAlbumBeanList.isNotEmpty()) {
                    picAlbumBeanList.forEach {
                        val existAlbum = picAlbumDao.getByInnerIndex(it.id)
                        if (existAlbum.exist != 1) {
                            picAlbumDao.update(it)
                            asyncStartDownload(
                                it.id,
                                picAlbumDataList.indexOf(picAlbumDataList.stream().filter { item ->
                                    item.picAlbumBean.id == it.id
                                }.findFirst().get())
                            );
                        }
                    }
                }
            }
        }
    }


    @SuppressLint("NotifyDataSetChanged")
    private val refreshFrontPage: () -> Unit = {
        picAlbumDataList.clear()
        val picAlbumBeanList = getDataSourceFromJsonFile()
        for (picAlbumBean in picAlbumBeanList) {
            val picAlbumData = PicAlbumData(picAlbumBean)
            picAlbumDataList.add(picAlbumData)
        }
        picAlbumListAdapter.notifyDataSetChanged()

    }

    private var isNotExistItemShown = true

    private fun getDataSourceFromJsonFile(): List<PicAlbumBean> {
        return if (isNotExistItemShown && NetworkUtil.isNetworkAvailable(this)) {
            picAlbumDao.getAll().toList()
        } else {
            picAlbumDao.getAllExist().toList()
        }
    }


    @SuppressLint("NotifyDataSetChanged")
    override fun doRefreshView(position: Int, currCount: Int, max: Int) {
        val viewHolder =
            listView.findViewHolderForAdapterPosition(position) as PicAlbumListAdapter.ViewHolder?
        if (currCount == max) {
            picAlbumDataList[position].picAlbumBean.exist = 1
            picAlbumDao.update(picAlbumDataList[position].picAlbumBean)
            picAlbumListAdapter.notifyDataSetChanged()
        }
        if (viewHolder != null) {
            MainScope().launch {
                viewHolder.downloadProcessBar.visibility = View.VISIBLE
                viewHolder.downloadProcessBar.isIndeterminate = false
                viewHolder.downloadProcessBar.setProgressCompat(currCount, false)
                viewHolder.downloadProcessBar.max = max
                Log.d(TAG, "current = $currCount max = $max")
            }
        }
    }


    fun asyncStartDownload(index: Long, position: Int) {
        downLoadService?.startDownloadAlbum(index, position)
    }

}